### PR TITLE
Use celery to send emails in background

### LINKED ISSUE #31: Bulk adding and emailing a ton of annotators is slow
There's a workaround: not adding way too many people at once (doing it in batches).

Not sure what's the best way to fix this.

### PR DESCRIPTION
**Pull request by evantey14**
This PR is to fix #31 
Still need to update the documentation and Procfile.

### CODE REVIEW

File: gavel/__init__.py [hunk 1]
```diff
@@ -23,6 +23,12 @@
 )
 assets.register('scss_all', scss)
 
+from celery import Celery
+app.config['CELERY_BROKER_URL'] = 'redis://localhost:6379/0'
+app.config['CELERY_RESULT_BACKEND'] = 'redis://localhost:6379/0'
```
Comment (anishathalye) [inline gavel/__init__.py]: Eventually, you should change this to read from `settings` (and then update the settings template and vagrant settings template accordingly). See line 10 in this file, for example.

Also, if these URLs are always the same (are they? idk), then you could just use a single setting for both.

File: gavel/utils.py [hunk 2]
```diff
@@ -85,7 +87,7 @@ def send_emails(emails):
             server.sendmail(settings.EMAIL_FROM, recipients, msg.as_string())
         except Exception as e:
             exceptions.append(e) # XXX is there a cleaner way to handle this?
-
+
```
Comment (anishathalye) [inline gavel/utils.py]: unnecessary change in the diff

File: requirements.txt [hunk 3]
```diff
@@ -15,10 +16,11 @@ psycopg2==2.7.1
...
+redis==2.10.5
 requests==2.13.0
 scipy==0.19.0
 six==1.10.0
 SQLAlchemy==1.1.6
 webassets==0.12.1
 Werkzeug==0.12.1
-xlrd==1.0.0
\ No newline at end of file
+xlrd==1.0.0
```
Comment (anishathalye) [inline requirements.txt]: why is this line part of the diff? (no newline at EOF?)

File: README.md [hunk 4]
```diff
@@ -73,6 +73,9 @@ When testing, you can run the app with `python runserver.py`. In production,
 you should use something like [Gunicorn][gunicorn] to serve this. You can run
```
Comment (anishathalye) [inline README.md]: Add in the paragraph above (that talks about postgres) that redis is also required if you want to send emails.

**Comment by anishathalye** [comment]
Oh also, at some point, we should test this on heroku. We'll probably need to update `app.json` so we can get a redis instance.

File: requirements.txt (hunk 3 above)
Comment (evantey14) [inline requirements.txt]: I'm not quite sure why this is here? I've been trying to play around and get rid of it, but it won't go away?

**Comment by evantey14** [comment]
Whoops I was hosed this week.
1. Nothing really should happen unless downloading redis or celery happens to fail. The send_email function won't get called, so nothing should get anywhere near touching redis/celery.
2. If celery is down, the broker will hold the tasks until a celery worker exists. Then it'll give those tasks to the worker  who will then send emails. I imagine if redis isn't there, it'll cause the same failure as if the database couldn't be found. I'll take a look into possibly falling back on sequential sending.

File: README.md [hunk 5]
```diff
@@ -65,14 +65,19 @@ install all the dependencies.
 
 The application uses Postgres for the database, so you need to have that on
 your server. You need to create a database, which you can do with `createdb
-gavel` (unless you're using a different database name). Before you use the app,
-you need to initialize the database by running `python initialize.py`. **Note
-that Gavel does not preserve database schema compatibility between versions.**
+gavel` (unless you're using a different database name). In order to send 
+emails, you'll need a broker on the server. You can do this using redis by
+running `redis-server`. Before you use the app,you need to initialize the
```
Comment (anishathalye) [inline README.md]: Just like we don't tell people how to install and run Postgres, I think we shouldn't explicitly tell people how to run Redis -- we should just tell them that they need to have Redis running. (Most likely, people won't be running it manually via `redis-server` -- they'll be using init scripts or something similar to run Redis as a service.)

File: requirements.txt (hunk 3 above)
Comment (anishathalye) [inline requirements.txt]: It's cause of a missing newline at EOF (usually, it's an editor setting whether or not you want a newline at the end of a file).

We use [EditorConfig](http://editorconfig.org/) in this project, so if you install EditorConfig for your editor, it'll automatically use the right conventions (due to this file / setting: https://github.com/anishathalye/gavel/blob/master/.editorconfig#L6).

File: gavel/settings.py [hunk 6]
```diff
@@ -58,6 +58,7 @@ def _list(item):
 BASE_URL =            c.get('base_url',        'BASE_URL')
 ADMIN_PASSWORD =      c.get('admin_password',  'ADMIN_PASSWORD')
 DB_URI =              c.get('db_uri',          ['DATABASE_URL', 'DB_URI'], default='postgresql://localhost/gavel')
+BROKER_URI =          c.get('broker_uri',      'BROKER_URI',               default='redis://localhost:6379/0')
```
Comment (anishathalye) [inline gavel/settings.py]: When Heroku supplies Redis, it provides you the URI via `REDIS_URL`. So you should use something like what's in line 60. Adapted for BROKER_URI, it would look like this: `BROKER_URI = c.get('broker_uri', ['REDIS_URL', 'BROKER_URI'], default='redis://localhost:6379/0')`

(and then also update the comment text in `config.template.yaml` to be like line 45 in that file, specifying both options for the env var)

**Comment by anishathalye** [comment]
If it's too much effort (or not possible) to automatically fall back to synchronously sending emails, we could add a `DISABLE_CELERY` option to the settings

**Comment by evantey14** [comment]
I'm personally more a fan of a `DISABLE_CELERY` option than automatic fallback to synchronous sending. I think at the moment, if the app can't find Redis, it'll error appropriately. If there's no celery worker, I think it makes more sense to error (or provide some feedback that celery can't be found)  than it does to fallback to synchronous sending? If I was trying to add 100 new judges, I think throwing an error seems nicer than attempting synchronous sending and failing at some indeterminate point along the way? (Let me know if I misunderstand how anything works.) What do you think?

For the feedback, yeah it seems like there isn't a great way to get celery task results after the user has loaded the page? Maybe something could be done so the user's page is reloaded only once the celery task is complete?

File: DEVELOPMENT.md [hunk 7]
```diff
@@ -69,6 +69,12 @@ Now, on your local machine, you should be able to navigate to
 `http://localhost:5000/admin` and login with the username "admin" with the
 password "admin".
 
+If you'd like to enable sending emails, you'll also need to run a celery worker
+with:
```
Comment (anishathalye) [inline DEVELOPMENT.md]: minor: add a newline between `with:` and the start of the ``` (it renders fine, but the plaintext is nicer to read if you have the newline, like the other examples).

File: README.md [hunk 8]
```diff
@@ -66,13 +66,19 @@ install all the dependencies.
 The application uses Postgres for the database, so you need to have that on
 your server. You need to create a database, which you can do with `createdb
 gavel` (unless you're using a different database name). Before you use the app,
-you need to initialize the database by running `python initialize.py`. **Note
+you need to initialize the database by running `python initialize.py`. **Note
```
Comment (anishathalye) [inline README.md]: minor: extraneous whitespace at EOL

File: README.md [hunk 9]
```diff
@@ -66,13 +66,19 @@ install all the dependencies.
 The application uses Postgres for the database, so you need to have that on
 your server. You need to create a database, which you can do with `createdb
 gavel` (unless you're using a different database name). Before you use the app,
-you need to initialize the database by running `python initialize.py`. **Note
+you need to initialize the database by running `python initialize.py`. **Note 
 that Gavel does not preserve database schema compatibility between versions.**
 
+In order to send emails, you'll need to have a broker, like Redis, to 
+distribute email tasks to a Celery worker.
```
Comment (anishathalye) [inline README.md]: Instead of saying that people need to install a broker like Redis, we should just tell people to install Redis (especially because the `requirements.txt` installs redis support, but doesn't install support for other brokers). Also, we don't need to explain the reason/mechanism here.

So maybe say something like "In order to send emails, you'll need to install Redis."

**Comment by evantey14** [comment]
Squashed.

**Comment by anishathalye** [comment]
Merged 1a4d362 in 2cdf0ab! 🎉

This patch has been merged into the `develop` branch. I'll merge `develop` into `master` once #61 is done (ideally, I'll get to it within this week).

Thanks for the contribution!
